layer-atlas/src/main/java/com/layer/atlas/util/imagepopup/AtlasImagePopupActivity.java

```diff
@@ -103,6 +129,65 @@ public static void init(LayerClient layerClient) {
         MessagePartRegionDecoder.init(layerClient);
     }
 
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.atlas_image_menu, menu);
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+       if(item.getItemId() == R.id.action_save) {
+           if(checkPermission()) {
+               saveImageToGallery();
+           } else {
+               requestPermission();
+           }
+           return true;
+       }
+       return super.onOptionsItemSelected(item);
+    }
+
+    private void saveImageToGallery() {
+        try {
+            String extractedID = UUID.randomUUID().toString();
+            List<String> path = mMessagePartId.getPathSegments();
+            if(path !=null && path.size()>2)
+                extractedID = path.get(path.size()-1);
+            MessagePartDecoder decoderFactory = new MessagePartDecoder();
+            Bitmap bitmap = decoderFactory.decode(this, mMessagePartId);
+            MediaStore.Images.Media.insertImage(
+                    getContentResolver(),
+                    bitmap,
+                    extractedID,
+                    ""
+            );
+            Toast.makeText(this, R.string.atlas_save_media_success, Toast.LENGTH_LONG).show();
```

Comment:
not true, insertImage can return null if image saving failed.

Reply:
Updated the logic for it but I'm not 100% happy with my implementation